lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,429 @@
+                         const char *Message,
+                         const char *File,
+                         int Line) {
+  if (Condition)
+    return;
+  dbg << "Warning: Assertion failed in " << File << ':' << Line << ":\n"
+      << What << "\n"
+      << Message << "\n\n";
+}
+
+#define soft_assert(what, message)                               \
+  if (Context.Options.SoftAsserts) {                             \
+    bool Condition = boolcast(what);                             \
+    print_assert(Condition, #what, message, __FILE__, __LINE__); \
+  } else {                                                       \
+    revng_check(what, message);                                  \
+  }
+
+// clang-format off
+template<typename T>
+using PlainStorage = std::conditional_t<anyOf<T, char *, const char *>(),
+                                        std::vector<char>,
+                                        T>;
+template<typename T>
+using ArgStorage = std::conditional_t<isList<T>(),
+                                      std::vector<std::remove_pointer_t<T>>,
+                                      PlainStorage<T>>;
+// clang-format on
+
+static llvm::ExitOnError AbortOnError;
+
+namespace utils {
+// Utility functions for decoding arguments
+
+// Put a nullptr at the end of the array and poison it, this will make it so
+// out-of-bounds errors are caught quickly
+template<typename T>
+inline void poisonTail(std::vector<T *> &Vector) {
+  Vector.push_back(nullptr);
+  ASAN_POISON_MEMORY_REGION(Vector.back(), sizeof(T *));
+}
+
+inline std::vector<const char *> toStringList(ArgumentRef Argument) {
+  std::vector<const char *> Result;
+  std::vector<llvm::StringRef> ArgList = Argument.toStringList();
+  for (llvm::StringRef Arg : ArgList) {
+    Result.push_back(Arg.data());
+  }
+  poisonTail(Result);
+  return Result;
+}
+
+} // namespace utils
+
+static std::vector<string> TemporaryDirectories;
+static void cleanTemporaryDirectories(void *) {
+  for (auto &Directory : TemporaryDirectories) {
+    std::filesystem::remove_all(Directory);
+  }
+}
+
+class RunnerContext {
+private:
+  llvm::StringMap<uintptr_t> Pointers;
+
+public:
+  const revng::tracing::RunTraceOptions Options;
+
+public:
+  RunnerContext(const revng::tracing::RunTraceOptions Options) :
+    Options(Options) {}
+  ~RunnerContext() = default;
+  RunnerContext(const RunnerContext &Other) = delete;
+  RunnerContext(RunnerContext &&Other) = delete;
+  RunnerContext &operator=(const RunnerContext &Other) = delete;
+  RunnerContext &operator=(RunnerContext &&Other) = delete;
+
+  const std::string getTemporaryDirectory() {
+    llvm::SmallString<128> TempOutput;
+    llvm::sys::fs::createUniqueDirectory("revng-run-trace", TempOutput);
+    std::string Output = TempOutput.str().str();
+    if (!Options.KeepTemporaries) {
+      TemporaryDirectories.push_back(Output);
+    } else {
+      std::cout << "Creating temporary directory: " << Output << "\n";
+      std::cout.flush();
```

Comment:
Why writing to `cout`?
Either use a logger or get a stream to write to in the constructor.
This looks like a logger.

Reply:
Swiched to `dbg`

Reply:
No, libraries cannot use `dbg` directly unless they abort immediately after.

Either use a logger or let `main` pass you `dbg`, so the main executable takes responsibility of letting you write to `dbg`.
But this looks like a logger.

Reply:
If a logger is used, then when you do `revng trace run --keep-temporaries` you're forced to add `--debug-log=trace-run` to actually see where the temporary folders are created, which is not that "ergonomic"

Reply:
1. Make it optional to override directories;
2. Have a `TemporaryDirectory` field in `Options` with a single temporary directory where you can create subdirectories;
3. Let `main` populate `Options.TemporaryDirectory` and print on `dbg` that it created it;

Writing to arbitrary file streams from a library is illegal!

Reply:
Changed the options